charms/worker/k8s/src/charm.py

```diff
@@ -149,10 +151,47 @@ def __init__(self, *args):
         )
 
         self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.feature_relation_changed, self._on_feature_relation_changed)
+
         if self.is_control_plane:
             self.etcd = EtcdReactiveRequires(self)
             self.framework.observe(self.on.get_kubeconfig_action, self._get_external_kubeconfig)
 
+    def _on_feature_relation_changed(self, event: ops.RelationChangedEvent):
+        relation_data = event.relation.data.get(event.unit) #type: ignore
+        if not relation_data:
+            log.warning("No relation data found for unit %s", event.unit)
+            return
+
```

Comment:
likely this will loop over all the feature relations here.

```python
for relation in self.model.relations["feature"]:
    for unit in relation.units:
        relation_data = relation.data[unit]
```